### PR TITLE
implemented xr.weighted, local tests pass

### PR DESCRIPTION
**Pull request by jbusecke**
I have implemented the use of `xr.weighted` in the `grid.average` function. I think this takes care of the issues mentioned in #193 and #182. 

Basically this now takes care of masking out weight values where the data is missing, to avoid the scenario in #182 , where we have a different number of valid values in each. 

Quick example:
```python
import xarray as xr
import numpy as np
from xgcm import Grid

# a minimal example
x = np.arange(10)
data= xr.DataArray(np.ones(10),dims='x', coords={'x':x})
weights = data * 30
ds = xr.Dataset({'data':data})
ds = ds.assign_coords(weights=weights)
ds
```
image

The current version behaves like this:
```python
# create an xgcm grid
grid = Grid(ds, coords={'X':{'center':'dim_0'}}, metrics={'X':['weights']})

# average
print(grid.average(ds.data, 'X'))

# now lets introduce a missing value in the data
data[6:8] = np.nan
# average
print(grid.average(ds.data, 'X'))
```
giving:
```
<xarray.DataArray ()>
array(1.)
<xarray.DataArray ()>
array(0.8)
```
This is obviously not right.

With the proposed changes the same code gives:
```
<xarray.DataArray ()>
array(1.)
<xarray.DataArray ()>
array(1.)
```
which is usually the expected output for such an operation.

It also happens to work with smaller subsets of data:
```python
# average
print(grid.average(ds.data[0:7], 'X'))
```
giving
```
<xarray.DataArray ()>
array(1.)
```
@miniufo, could you check if this addresses your issue in #193 ?

We should still have a broader conversation about how to deal with subsets, but personally I think `grid.average` is often used after the other functions are applied to a global dataset (at least in my personal experience). So having the ability to e.g. calculate the x-gradient and then select a box and average it only over that box, is already a big improvement. 

Thoughts?

### CODE REVIEW

**Comment by miniufo** [comment]
>@miniufo, could you check if this addresses your issue in #193 ?

Glad you work this out and it is a good start.  But I'm a little confused.  Do I need to explicitly declare a new DataArray `weights`? Or just convert a field (e.g., SST) into `weights` (0 when SST is nan and 1 otherwise)?  I am expecting that `xgcm` will do this for users (hopeless lazy guy here...)

Do I have to update the whole `xgcm` or just modified the `grid.average` function as in this commit to take effect?  I am not so sure how to test the new codes when it is in a branch that is not merged into master.

**Comment by jbusecke** [comment]
You should not have to do anything else than pulling/installing this pull request. In a conda environment you can just do `pip install git+https://github.com/xgcm/xgcm.git@refs/pull/204/head` (make sure this is a separate env from your main one, so that you can ditch it, if things fail).

Once that is done, execute your example from #193 again, and hopefully it just works correctly. No additional input should be required from your side. Thanks for the the help.

**Comment by dcherian** [comment]
Alternatively if you have an editable install of xgcm (`pip install -e`), you can do `gh pr checkout 204` in the xgcm root directory (https://cli.github.com/)

File: xgcm/grid.py [hunk 1]
```diff
@@ -1431,24 +1431,21 @@ def average(self, da, axis, **kwargs):
...
+            Additional arguments passed to `xarray.DataArray.weighted.mean`
 
 
         Returns
         -------
         da_i : xarray.DataArray
             The averaged data
         """
-
+        da = da.copy()
```
Comment (rabernat) [inline xgcm/grid.py:1442]: why do we need to copy?

Comment (jbusecke) [reply xgcm/grid.py:1442]: Hmm probably we dont.

**Comment by jbusecke** [comment]
One thing I noticed just now, is that we probably need at least a warning if the alignment results in an empty array. I was just beating my head against the wall until I realized I was using a grid from another model. Just dropping it here, so I dont forget.

**Comment by jbusecke** [comment]
#111  is relevant to this discussion aswell. 

I think either way we need some validation of the input dimensions.

**Comment by jbusecke** [comment]
I messed this git history up beyond repair (at least for me). Decided to start fresh in #259 and closing this.

**Pull request closed without merging**